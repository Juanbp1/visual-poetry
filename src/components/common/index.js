/**
 * @module common
 * @description Biblioteca de componentes de interfaz de usuario. Proporciona una serie de componentes reutilizables
 * para la construcción de interfaces de usuario dinámicas y responsivas.
 * 
 * @see {@link module:common/Button|Button} - Componente de botón reutilizable con varias propiedades de estilo y eventos.
 * @see {@link module:common/Logo|Logo} - Componente para mostrar el logotipo.
 * @see {@link module:common/ResponsiveComponent|ResponsiveComponent} - Componente que ajusta su renderizado según el tamaño de la pantalla.
 * @see {@link module:common/Textarea|Textarea} - Componente de área de texto con características avanzadas.
 */
export { default as Button } from './Button';
export { default as Logo } from './Logo';
export { default as ResponsiveComponent } from './ResponsiveComponent';
export { default as Textarea } from './Textarea';
