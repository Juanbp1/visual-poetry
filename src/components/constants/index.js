export { default as actions } from './actions';
export { default as actionsFactory } from './actionsFactory';
export { default as actionTypes } from './actionTypes';
// export { default as FAQS_DATA } from './faqsData';
// export { default as SOCIAL_NETWORKS } from './socialNetworks';
// export { default as STEPS_DATA } from './stepsData';
// export { default as uiConstants } from './uiConstants';
