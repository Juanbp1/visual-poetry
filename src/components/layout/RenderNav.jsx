import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
/**
 * Componente que renderiza un menú de navegación.
 *
 * @module layout/RenderNav
 * @param {Object} props - Propiedades del componente.
 * @param {Array<Object>} props.menuItems - Array de objetos que representan los elementos del menú.
 * @param {string} props.menuItems[].href - URL a la que debe enlazar el elemento del menú.
 * @param {string} props.menuItems[].label - Texto del enlace del menú.
 * @param {string} props.className - Clase adicional que se aplica a los elementos de navegación.
 * @returns {JSX.Element} El componente de nav.
 * 
 * @example
 * import React from 'react';
 * import { RenderNav } from '../layout'
 * const menuItems = [
 *   { href: '/home', label: 'Home' },
 *   { href: '/about', label: 'About' },
 *   { href: '/contact', label: 'Contact' },
 * ];
 *
 * const MyComponent = () => {
 *  return(
 *     <RenderNav
 *       menuItems={menuItems}
 *       className="main-menu"
 *     />
 *   );
 *  }
 * export default MyComponent;
 */
const RenderNav = ({ menuItems, className }) => {
  const getClassName = (baseClass) => classNames(baseClass, {
    [`${baseClass}--${className}`]: className
  });
  
  const headerNav = getClassName('header__nav');
  const headerUl = getClassName('header__ul');
  const headerLi = getClassName('header__li');
  const headerLink = getClassName('header__link');
  return (
    <nav className={headerNav} data-testid={`${className}`}>
      <ul className={headerUl}>
        {menuItems.map((link, index) => (
          <li key={index} className={headerLi}>
            <a
              className={headerLink}
              href={link.href}
              data-testid={`href--${className}`}
            >
              {link.label}
            </a>
          </li>
        ))}
      </ul>
    </nav>
  ); 
};
RenderNav.propTypes = {
  menuItems: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  className: PropTypes.string.isRequired,
};
export default RenderNav;
