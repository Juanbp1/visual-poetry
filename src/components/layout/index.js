/**
 * @module layout
 * @description Conjunto de componentes principales de la aplicación. Incluye elementos esenciales para la estructura de la interfaz, como el pie de página, la cabecera, la barra lateral y más.
 * 
 * @see {@link module:layout/Footer|Footer} - Componente que representa el pie de página de la aplicación.
 * @see {@link module:layout/header/Header|Header} - Componente que representa el header de la aplicación.
 * @see {@link module:layout/header/DesktopHeader|DesktopHeader} - Componente que representa el header para dispositivos laptop.
 * @see {@link module:layout/header/MobileHeader|MobileHeader} - Componente que representa el header para dispositivos movil.
 * @see {@link module:layout/RenderNav|RenderNav} - Componente que renderiza un menú de navegación.
 * @see {@link module:layout/SideBar|SideBar} - Componente que representa la barra lateral de la aplicación.
 * @see {@link module:layout/SocialNetworks|SocialNetworks} - Componente para mostrar los enlaces a redes sociales.
 * @see {@link module:layout/EditorToolbox|EditorToolbox} - Componente que proporciona herramientas para la edición de texto.
 */

export { default as Footer } from './Footer';
export { default as Header } from './header/Header';
export { default as MobileHeader } from './header/MobileHeader';
export { default as DesktopHeader } from './header/DesktopHeader';
export { default as RenderNav } from './RenderNav';
export { default as SideBar } from './SideBar';
export { default as SocialNetworks } from './SocialNetworks';
export { default as EditorToolbox } from './EditorToolbox';
