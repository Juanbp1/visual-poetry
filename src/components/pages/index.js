/**
 * @module pages
 * 
 * @description Este módulo exporta los componentes principales utilizados en la aplicación. Cada componente tiene un propósito específico y contribuye a la funcionalidad y la interfaz de usuario de la aplicación.
 * 
 * @see {@link module:pages/Home|Home} - Componente principal que representa la página de inicio de la aplicación.
 * @see {@link module:pages/AppViewer|AppViewer} - Componente que proporciona una vista detallada de la aplicación, permitiendo la interacción con sus características principales.
 * @see {@link module:pages/NotFound|NotFound} - Componente que se muestra cuando el usuario accede a una página no encontrada o inexistente.
 * 
 */
export { default as Home } from './Home';
export { default as AppViewer } from './AppViewer';
export { default as NotFound } from './NotFound';
