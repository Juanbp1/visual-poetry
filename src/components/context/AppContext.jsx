/**
 * @fileoverview Contexto y Reducer para la aplicación.
 * @module AppContext
 */
import React, { createContext, useContext, useReducer } from 'react';
import actionTypes from '../constants/actionTypes';
import { TEXT_STRINGS } from '../constants/uiConstants';
import actionsFactory from '../constants/actionsFactory';
import defineActions from '../constants/actions';

/**
 * @typedef {Object} AppState
 * @property {Image|null} image - La imagen actual.
 * @property {ImageBorder|null} imageBorder - El borde de la imagen.
 * @property {boolean} imageIsInverted - Indica si la imagen está invertida.
 * @property {Array<ImageGalleryItem>} imageGallery - La galería de imágenes.
 * @property {Object|null} imageSelected - La imagen seleccionada.
 * @property {boolean} imageIsLoaded - Indica si la imagen ha sido cargada.
 * @property {boolean} imageIsProcessed - Indica si la imagen ha sido procesada.
 * @property {boolean} inputIsSetted - Indica si la entrada está configurada.
 * @property {Object} imageBorderPixels - Los píxeles del borde de la imagen.
 * @property {Object} imageData - Datos de la imagen.
 * @property {Object} inputPositionCoordinates - Coordenadas de posición de entrada.
 * @property {string} inputData - Datos de entrada.
 * @property {Object} mainContext - Contexto principal del canvas.
 * @property {Object} secondContext - Segundo contexto del canvas.
 * @property {Object} canvasSize - Tamaño del canvas.
 * @property {Object} styleFont - Estilo de fuente actual.
 * @property {string|null} errorUploadImage - Error al subir la imagen.
 * @property {number} figureThreshold - Umbral para figuras.
 * @property {boolean} loading - Estado de carga.
 * @property {boolean} showGallery - Indica si se muestra la galería.
 * @property {boolean} showGallerySecond - Indica si se muestra la segunda galería.
 * @property {boolean} showGalleryThird - Indica si se muestra la tercera galería.
 */

/**
 * @type {React.Context<{state: AppState, dispatch: React.Dispatch<any>, actions: Object}>}
 */
const AppContext = createContext();

/**
 * Estado inicial del contexto de la aplicación.
 * @type {AppState}
 */
const initialState = {
  image: null,
  imageBorder: null,
  imageIsInverted: false,
  imageGallery: [],
  imageSelected: null,
  imageIsLoaded: false,
  imageIsProcessed: false,
  inputIsSetted: false,
  imageBorderPixels: {},
  imageData: {},
  inputPositionCoordinates: { start: [], end: [] },
  inputData: TEXT_STRINGS.INITIAL_TEXT,
  mainContext: null,
  secondContext: null,
  canvasSize: { canvasWidth: 0, canvasHeight: 0 },
  styleFont: {
    color: '#000000',
    background: '#FF000000',
    font: 'Arial',
    size: 8,
  },
  errorUploadImage: null,
  figureThreshold: 128,
  loading: false,
  showGallery: false,
  showGallerySecond: false,
  showGalleryThird: false,
};

const {
  CHECK_IMAGE_LOADED,
  CHECK_IMAGE_PROCESSED,
  CHECK_INPUT_SETTED,
  OBTAIN_CANVAS_SIZE,
  OBTAIN_IMAGE,
  OBTAIN_IMAGE_BORDER,
  OBTAIN_IMAGE_BORDER_PIXELS,
  OBTAIN_IMAGE_GALLERY,
  OBTAIN_IMAGE_SELECTED,
  OBTAIN_INPUT_POSITION_COORDINATES,
  OBTAIN_MAIN_CONTEXT,
  OBTAIN_SECOND_CONTEXT,
  RESET,
  TOGGLE_INVERT_IMAGE,
  UPDATE_ERROR_UPLOAD_IMAGE,
  UPDATE_FIGURE_THRESHOLD,
  UPDATE_GALLERY_TOGGLE,
  UPDATE_GALLERY_TOGGLE2,
  UPDATE_GALLERY_TOGGLE3,
  UPDATE_IMAGE_GALLERY,
  UPDATE_INPUT_DATA,
  UPDATE_LOADING,
  UPDATE_STYLE_FONT,
} = actionTypes;

/**
 * Función reductora para manejar acciones en el estado de la aplicación.
 * @param {AppState} state - El estado actual de la aplicación.
 * @param {Object} action - La acción que se va a procesar.
 * @param {string} action.type - El tipo de acción que indica qué operación realizar.
 * @param {any} [action.payload] - Los datos asociados a la acción, si los hay.
 * @returns {AppState} - El nuevo estado de la aplicación después de procesar la acción.
 */
const reducer = (state, action) => {
  switch (action.type) {
    case CHECK_IMAGE_LOADED:
      return { ...state, imageIsLoaded: true };

    case CHECK_IMAGE_PROCESSED:
      return { ...state, imageIsProcessed: true };

    case CHECK_INPUT_SETTED:
      return { ...state, inputIsSetted: true };

    case OBTAIN_CANVAS_SIZE:
      return { ...state, canvasSize: action.payload };

    case OBTAIN_IMAGE:
      return { ...state, image: action.payload };

    case OBTAIN_IMAGE_BORDER:
      return { ...state, imageBorder: action.payload };

    case OBTAIN_IMAGE_BORDER_PIXELS:
      if (action.payload && typeof action.payload.getImageData === 'function') {
        const context = action.payload;
        const imageDataPixels = context.getImageData(
          0,
          0,
          context.canvas.width,
          context.canvas.height
        );
        return { ...state, imageBorderPixels: imageDataPixels };
      }
      return { ...state, imageBorderPixels: {} };
    case OBTAIN_IMAGE_GALLERY:
      return { ...state, imageGallery: action.payload };

    case OBTAIN_INPUT_POSITION_COORDINATES:
      return { ...state, inputPositionCoordinates: action.payload };

    case OBTAIN_MAIN_CONTEXT:
      return { ...state, mainContext: action.payload };

    case OBTAIN_SECOND_CONTEXT:
      return { ...state, secondContext: action.payload };

    case OBTAIN_IMAGE_SELECTED:
      return { ...state, imageSelected: action.payload };

    case RESET:
      return { ...initialState };

    case TOGGLE_INVERT_IMAGE:
      return { ...state, imageIsInverted: action.payload };

    case UPDATE_ERROR_UPLOAD_IMAGE:
      return { ...state, errorUploadImage: action.payload };

    case UPDATE_FIGURE_THRESHOLD:
      return { ...state, figureThreshold: action.payload };

    case UPDATE_GALLERY_TOGGLE:
      return { ...state, showGallery: action.payload };

    case UPDATE_GALLERY_TOGGLE2:
      return { ...state, showGallerySecond: action.payload };

    case UPDATE_GALLERY_TOGGLE3:
      return { ...state, showGalleryThird: action.payload };

    case UPDATE_IMAGE_GALLERY:
      const id = action.payload;
      const imageSelected = state.imageGallery[0].map((img) => {
        return { ...img, selected: img.id === id ? !img.selected : false };
      });
      return { ...state, imageGallery: [imageSelected] };

    case UPDATE_INPUT_DATA:
      return { ...state, inputData: action.payload };

    case UPDATE_LOADING:
      return { ...state, loading: !state.loading };

    case UPDATE_STYLE_FONT:
      const updatedStyleFont = {
        ...state.styleFont,
        ...action.payload,
      };
      return {
        ...state,
        styleFont: updatedStyleFont,
      };

    default:
      return state;
  }
};

/**
 * Proveedor del contexto de la aplicación.
 * @param {Object} props - Las propiedades del componente.
 * @param {React.ReactNode} props.children - Los componentes hijos a envolver.
 * @returns {JSX.Element} - El componente Provider con el contexto.
 */
export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const createAction = actionsFactory(dispatch);

  // Definir las acciones utilizando la fábrica de acciones
  const { imageActions, styleActions, inputActions, othersActions } =
    defineActions(createAction);

  const actions = {
    ...imageActions,
    ...styleActions,
    ...inputActions,
    ...othersActions,
  };

  return (
    <AppContext.Provider value={{ state, dispatch, actions }}>
      {children}
    </AppContext.Provider>
  );
};

/**
 * Hook para utilizar el contexto de la aplicación.
 * @returns {Object} - El contexto de la aplicación que incluye el estado, el despachador y las acciones.
 * @throws {Error} - Lanza un error si el hook se usa fuera del proveedor de contexto.
 */
export const useAppContext = () => {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error(
      'useAppContext debe ser utilizado dentro de un AppProvider'
    );
  }
  return context;
};
