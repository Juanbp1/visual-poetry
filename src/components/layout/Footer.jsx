import React from 'react';
import { scrollToTop } from '../utils';
import { IoArrowUp } from 'react-icons/io5';
import { SocialNetworks } from './';
import { Logo, Button, ResponsiveComponent } from '../common';
import { SOCIAL_NETWORKS } from '../constants/socialNetworks';
import { DEVICE_TYPES } from '../constants/uiConstants';
const { MOBILE, TABLET, LAPTOP } = DEVICE_TYPES;
/**
 *Componente que representa el pie de página de la aplicación.
 * @module layout/Footer
 * @returns {JSX.Element} El componente Footer
 * @example
 * import React from 'react';
 * import { Footer } from '../layout'
 *
 * const MyComponent = () => {
 *   <Footer />
 * };
 * export default MyComponent;
 */
const Footer = () => {
  return (
    <section className="sections__footerContainer footerContainer">
      <div className="footerContainer__footer footer">
        <div className="footer__content">
          <div className="uiLogo__logo">
            <Logo className="footer" />
          </div>
          <ResponsiveComponent
            visibleOnDeviceTypes={[TABLET, LAPTOP]}
            componentToRender={
              <Button
                className="button__footer button__withIcon "
                name="Volver arriba"
                text="Volver arriba"
                handleClick={() => scrollToTop()}
                Icon={IoArrowUp}
                iconClassName="button__footerArrowIcon"
                title="Volver al Inicio"
                ariaLabel="Volver al Inicio"
                id="footerButton"
              />
            }
          />
          <ResponsiveComponent
            visibleOnDeviceTypes={[MOBILE]}
            componentToRender={
              <Button
                className="button__footer button__withIcon button__circle"
                handleClick={() => scrollToTop()}
                Icon={IoArrowUp}
                iconClassName="button__footerArrowIcon"
                title="Volver al Inicio"
                ariaLabel="Volver al Inicio"
                id="footerButton"
              />
            }
          />
        </div>
        <div className="socialMedia__socialCopyManager">
          <SocialNetworks
            socialNetworks={SOCIAL_NETWORKS}
            classContainer="footer"
          />
          <p className="socialMedia__copyright ">
            Copyright © 2024 Juan Manuel Barea Pantoja
          </p>
        </div>
      </div>
    </section>
  );
};

export default Footer;
