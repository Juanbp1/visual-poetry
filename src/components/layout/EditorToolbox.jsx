import React, { useCallback } from 'react';
import { useAppContext } from '../context/AppContext';
import { FaCircle } from 'react-icons/fa';
import { IoInvertMode } from 'react-icons/io5';
import { SIZE_TEXT, FONTS } from '../constants/uiConstants';
import {
  SelectionTool,
  CanvasTextEditor,
  ColorPickerControl,
  SelectDropdown,
  Slider,
} from '../VisualPoetry/canvasToolbox';

/**
 * Componente que representa la caja de herramientas del canvas.
 * @module layout/EditorToolbox
 * @returns {JSX.Element} El componente de la caja de herramientas del canvas.
 * @example
 * // Uso del componente EditorToolbox
 * import React from 'react';
 * import { EditorToolbox } from '../layout';
 *
 * const MyComponent = () => {
 * return(
 *   <EditorToolbox/>
 *  );
 * }
 * export default MyComponent;
 */
const EditorToolbox = () => {
  const { state, actions } = useAppContext();
  const { styleFont } = state;
  const { background, color, font } = styleFont;
  const {
    updateStyleColorAction,
    updateStyleBackgroundColorAction,
    updateStyleFontAction,
    updateStyleSizeAction,
  } = actions;

  /**
   * Maneja el cambio de color del texto.
   * @member module:layout/EditorToolbox
   * @param {string} newColor - El nuevo color que se aplicará al texto.
   */
  const handleInputColor = (newColor) => {
    updateStyleColorAction({ color: newColor });
  };

  /**
   * Maneja el cambio de color de fondo del lienzo.
   * @member module:layout/EditorToolbox
   * @param {string} newBackground - El nuevo color de fondo que se aplicará al lienzo.
   */
  const handleInputBackgroundColor = (newBackground) => {
    updateStyleBackgroundColorAction({ background: newBackground });
  };

  /**
   * Maneja la selección de la familia de fuentes para el texto.
   * @member module:layout/EditorToolbox
   * @param {React.ChangeEvent<HTMLSelectElement>} event - El evento de cambio desencadenado al seleccionar una nueva familia de fuentes.
   */
  const handleSelectFontFamily = useCallback(
    (event) => {
      const newFont = `${event.target.value}`;
      updateStyleFontAction({ font: newFont });
    },
    [updateStyleFontAction]
  );

  /**
   * Maneja la selección del tamaño del texto.
   * @member module:layout/EditorToolbox
   * @param {React.ChangeEvent<HTMLSelectElement>} event - El evento de cambio se activa al seleccionar un nuevo tamaño de texto.
   */
  const handleSelectSize = useCallback(
    (event) => {
      const newSize = `${event.target.value}`;
      updateStyleSizeAction({ size: newSize });
    },
    [updateStyleSizeAction]
  );

  return (
    <div className="toolbox" data-testid="toolbox">
      <ColorPickerControl
        id="color"
        IconComponent={FaCircle}
        currentColor={color}
        handleChange={handleInputColor}
        title="Color del texto"
      />
      <ColorPickerControl
        id="background"
        IconComponent={FaCircle}
        currentColor={background}
        handleChange={handleInputBackgroundColor}
        title="Color del fondo"
      />
      <SelectDropdown
        id="fonts"
        handleChange={handleSelectFontFamily}
        title="Fuente del texto"
        options={FONTS}
        defaultValue={font}
      />
      <SelectDropdown
        id="size"
        handleChange={handleSelectSize}
        title="Tamaño de texto"
        options={SIZE_TEXT}
        defaultValue={styleFont.size}
      />
      <SelectionTool
        id="selectionTool"
        IconComponent={IoInvertMode}
        title="Herramienta selección"
      />
      <Slider id="cutPoint" title="Punto de corte" />
      <CanvasTextEditor title="Texto" id="textEditor" />
    </div>
  );
};

export default EditorToolbox;
