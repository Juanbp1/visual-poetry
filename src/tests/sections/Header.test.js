import React from 'react';
import { cleanup, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router';
import userEvent from '@testing-library/user-event';
import { AppProvider } from '../../components/context/AppContext';
import { Home, AppViewer } from '../../components/pages';
import { mockUseMediaQuery } from '../mockUseMediaQuery';

import { verifyLinks } from '../helpers/userHelpers';
import { Header } from '../../components/layout';
import { RenderNav } from '../../components/layout';

const MENU = [
  { href: '#quickStartId', label: 'Inicio rápido' },
  { href: '#dropzoneId', label: 'Subir imagen' },
  { href: '#exampleGalleryId', label: 'Galeria de ejemplos' },
  { href: '#faqsId', label: 'Preguntas frecuentes' },
];

const renderDesktopHeader = () => {
  return render(
    <AppProvider>
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            exact
            path="/"
            element={
              <Home>
                <RenderNav menuItems={MENU} className="--landscape" />
              </Home>
            }
          />
        </Routes>
      </MemoryRouter>
    </AppProvider>
  );
};

const renderMobileHeader = () => {
  mockUseMediaQuery({ isMobile: true });
  return render(
    <AppProvider>
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route
            path="/"
            element={
              <Home>
                <Header
                  classContainer="frontPageContainer__header"
                  isOpenSideBar={true}
                />
              </Home>
            }
          />
        </Routes>
      </MemoryRouter>
    </AppProvider>
  );
};

describe('El header debe funcionar correctamente (desktop)', () => {
  afterEach(() => {
    cleanup();
  });
  // Verifica si el logo del header se renderiza correctamente comparando con un snapshot
  test('El logo se muestra correctamente', () => {
    renderDesktopHeader();
    const logo = screen.getByTestId('headerLogo');
    expect(logo).toMatchSnapshot();
  });

  // Verifica si el elemento de navegación del header se renderiza con las clases esperadas
  test('El nav de navegación se muestra correctamente', async () => {
    renderDesktopHeader();

    const headerNav = screen.getByTestId('landscape');
    expect(headerNav).toHaveClass('header__nav header__nav--landscape');
    expect(headerNav).toMatchSnapshot();
  });

  // Verifica que los enlaces de navegación redirigen correctamente a las secciones correspondientes
  test('Debería desplazarse a la sección correspondiente cuando se hace clic en un enlace de navegación', async () => {
    renderDesktopHeader();

    const links = screen.getAllByTestId('href--landscape');
    verifyLinks(links, MENU);
  });

  // Verifica si la clase header--scrolled se añade o se elimina correctamente según la posición de desplazamiento de la ventana.
  test('Debería cambiar la class de scroll según la posición de desplazamiento', async () => {
    renderDesktopHeader();
    const headerElement = screen.getByRole('banner');

    // Simula un scroll hacia abajo en la página
    window.scrollY = 150;
    window.dispatchEvent(new Event('scroll'));

    // Verifica si la class de scrolled se añade correctamente
    await waitFor(() => {
      expect(headerElement.className).toContain('header--scrolled');
    });

    // Simula un scroll hacia arriba en la página
    window.scrollY = 0; // Simulate scrolling up
    window.dispatchEvent(new Event('scroll'));

    // Verifica si la class de scrolled se elimina correctamente
    await waitFor(() => {
      expect(headerElement.className).not.toContain('header--scrolled');
    });
  });

  // Verficia que la función handleCloseGallery no se ejecuta cuando la aplicación está en modo móvil y ocurre un evento de redimensionamiento (resize).
  test('handleCloseGallery no debería llamarse cuando no es laptop y ocurre resize', async () => {
    const handleCloseGalleryMock = jest.fn();
    renderDesktopHeader();

    // Simular condición de vista móvil
    mockUseMediaQuery({ isMobile: true });

    // Simula un evento de redimensionamiento de ventana
    await waitFor(() => {
      window.dispatchEvent(new Event('resize'));
    });

    // Verifica que la función handleCloseGallery no se llama cuando se redimensiona la ventana en vista móvil
    expect(handleCloseGalleryMock).not.toHaveBeenCalled();
  });
});

describe('El header debe funcionar correctamente (mobile)', () => {
  afterEach(() => {
    cleanup();
  });
  test('Al hacer clic el botón de menú se debería abrir el sidebar', async () => {
    mockUseMediaQuery({ isMobile: true });

    renderMobileHeader();
    const menuButton = screen.getByLabelText('Abrir menu de navegación');
    const sidebar = await screen.findByTestId('menuSidebar');

    userEvent.click(menuButton);

    expect(sidebar).toBeVisible();
  });
  test('Al hacer clic el botón de close se debería cerrar el sidebar', async () => {
    mockUseMediaQuery({ isMobile: true });

    renderMobileHeader();
    const menuButton = screen.getByLabelText('Abrir menu de navegación');
    const closeButon = screen.getByTestId('menuSidebarClose');
    const sidebar = await screen.findByTestId('menuSidebar');

    userEvent.click(menuButton);
    expect(sidebar).toBeVisible();

    // El sidebar debe estar abierto
    await waitFor(() => {
      expect(sidebar).toHaveStyle('transform: translateX(0%);');
    });

    userEvent.click(closeButon);

    // El sidebar debe estar cerrado
    await waitFor(() => {
      expect(sidebar).toHaveStyle('transform: translateX(100%);');
    });
  });
  test('Debería desplazarse a la sección correspondiente cuando se hace clic en un enlace de navegación', async () => {
    mockUseMediaQuery({ isMobile: true });

    renderMobileHeader();
    const menuButton = screen.getByLabelText('Abrir menu de navegación');
    const sidebar = await screen.findByTestId('menuSidebar');

    userEvent.click(menuButton);
    expect(sidebar).toBeVisible();

    // El sidebar debe estar abierto
    await waitFor(() => {
      expect(sidebar).toHaveStyle('transform: translateX(0%);');
    });

    const links = screen.getAllByTestId('href--portrait');
    verifyLinks(links, MENU);
  });
  test('El logo se muestra correctamente', () => {
    renderMobileHeader();

    const logo = screen.getByTestId('uiLogoLogo');
    expect(logo).toMatchSnapshot();
  });
});
