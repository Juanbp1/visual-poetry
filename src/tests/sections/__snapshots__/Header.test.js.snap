// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`El header debe funcionar correctamente (desktop) El logo se muestra correctamente 1`] = `
<a
  class="header__link--notUnderlined"
  data-testid="headerLogo"
  href="/"
>
  <span
    aria-label="logo de la p치gina web"
    class="uiLogo__logo"
  >
    <svg
      class="uiLogo__logoIcon uiLogo__logoIcon--header"
      version="1.1"
      viewBox="32798 1839 45 53.598"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        id="shape"
      >
        <g
          class="fills"
        >
          <path
            class="uiLogo__logoIcon uiLogo__logoIcon--header"
            d="M32819.572,1839.016C32820.318,1839.007,32821.079,1838.981,32821.822,1839.025C32832.019,1839.625,32840.751,1847.626,32842.625,1857.636C32843.863,1864.248,32842.012,1871.414,32837.387,1876.360C32835.487,1878.392,32834.341,1878.856,32834.343,1881.772L32834.345,1884.584C32834.345,1886.148,32834.263,1887.604,32832.443,1888.215L32832.377,1888.236C32831.957,1888.375,32831.390,1888.400,32830.963,1888.285L32830.906,1888.269C32828.714,1887.668,32828.814,1885.930,32828.813,1884.174L32828.812,1881.680C32828.811,1881.243,32828.853,1880.698,32828.725,1880.278C32827.897,1877.566,32824.267,1877.384,32823.411,1880.062C32823.168,1880.821,32823.273,1881.764,32823.273,1882.558L32823.277,1887.654C32823.276,1888.809,32823.482,1890.056,32822.962,1891.126C32822.363,1892.182,32821.216,1892.749,32820.003,1892.563L32819.947,1892.554C32817.906,1892.236,32817.681,1890.620,32817.680,1888.992L32817.679,1876.255C32817.678,1875.918,32817.712,1875.524,32817.616,1875.199C32816.823,1872.507,32813.213,1872.167,32812.372,1874.895L32812.352,1874.960C32811.917,1876.365,32812.772,1877.977,32811.631,1879.286C32810.101,1881.042,32808.196,1880.243,32806.677,1879.073L32805.299,1877.949C32791.627,1865.971,32798.290,1843.402,32815.924,1839.497C32817.126,1839.231,32818.345,1839.092,32819.572,1839.016ZZM32820.144,1853.839C32818.307,1854.000,32816.672,1854.768,32815.484,1856.200C32811.897,1860.525,32815.243,1867.030,32820.941,1866.720C32826.631,1866.328,32829.122,1859.154,32824.643,1855.367L32823.680,1854.684C32822.718,1854.097,32821.294,1853.672,32820.144,1853.839ZZ"
            fill="black"
          />
        </g>
      </g>
    </svg>
    <p
      class="uiLogo__logoName uiLogo__logoName--header"
    >
      Visual Poetry
    </p>
  </span>
</a>
`;

exports[`El header debe funcionar correctamente (desktop) El nav de navegaci칩n se muestra correctamente 1`] = `
<nav
  class="header__nav header__nav--landscape"
  data-testid="landscape"
>
  <ul
    class="header__ul header__ul--landscape"
  >
    <li
      class="header__li header__li--landscape"
    >
      <a
        class="header__link header__link--landscape"
        data-testid="href--landscape"
        href="/#quickStartId"
      >
        Inicio r치pido
      </a>
    </li>
    <li
      class="header__li header__li--landscape"
    >
      <a
        class="header__link header__link--landscape"
        data-testid="href--landscape"
        href="/#dropzoneId"
      >
        Subir imagen
      </a>
    </li>
    <li
      class="header__li header__li--landscape"
    >
      <a
        class="header__link header__link--landscape"
        data-testid="href--landscape"
        href="/#exampleGalleryId"
      >
        Galeria de ejemplos
      </a>
    </li>
    <li
      class="header__li header__li--landscape"
    >
      <a
        class="header__link header__link--landscape"
        data-testid="href--landscape"
        href="/#faqsId"
      >
        Preguntas frecuentes
      </a>
    </li>
  </ul>
</nav>
`;

exports[`El header debe funcionar correctamente (mobile) El logo se muestra correctamente 1`] = `
<a
  class="header__link--notUnderlined"
  data-testid="uiLogoLogo"
  href="/"
>
  <span
    aria-label="logo de la p치gina web"
    class="uiLogo__logo"
  >
    <svg
      class="uiLogo__logoIcon uiLogo__logoIcon--uiLogo"
      version="1.1"
      viewBox="32798 1839 45 53.598"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        id="shape"
      >
        <g
          class="fills"
        >
          <path
            class="uiLogo__logoIcon uiLogo__logoIcon--uiLogo"
            d="M32819.572,1839.016C32820.318,1839.007,32821.079,1838.981,32821.822,1839.025C32832.019,1839.625,32840.751,1847.626,32842.625,1857.636C32843.863,1864.248,32842.012,1871.414,32837.387,1876.360C32835.487,1878.392,32834.341,1878.856,32834.343,1881.772L32834.345,1884.584C32834.345,1886.148,32834.263,1887.604,32832.443,1888.215L32832.377,1888.236C32831.957,1888.375,32831.390,1888.400,32830.963,1888.285L32830.906,1888.269C32828.714,1887.668,32828.814,1885.930,32828.813,1884.174L32828.812,1881.680C32828.811,1881.243,32828.853,1880.698,32828.725,1880.278C32827.897,1877.566,32824.267,1877.384,32823.411,1880.062C32823.168,1880.821,32823.273,1881.764,32823.273,1882.558L32823.277,1887.654C32823.276,1888.809,32823.482,1890.056,32822.962,1891.126C32822.363,1892.182,32821.216,1892.749,32820.003,1892.563L32819.947,1892.554C32817.906,1892.236,32817.681,1890.620,32817.680,1888.992L32817.679,1876.255C32817.678,1875.918,32817.712,1875.524,32817.616,1875.199C32816.823,1872.507,32813.213,1872.167,32812.372,1874.895L32812.352,1874.960C32811.917,1876.365,32812.772,1877.977,32811.631,1879.286C32810.101,1881.042,32808.196,1880.243,32806.677,1879.073L32805.299,1877.949C32791.627,1865.971,32798.290,1843.402,32815.924,1839.497C32817.126,1839.231,32818.345,1839.092,32819.572,1839.016ZZM32820.144,1853.839C32818.307,1854.000,32816.672,1854.768,32815.484,1856.200C32811.897,1860.525,32815.243,1867.030,32820.941,1866.720C32826.631,1866.328,32829.122,1859.154,32824.643,1855.367L32823.680,1854.684C32822.718,1854.097,32821.294,1853.672,32820.144,1853.839ZZ"
            fill="black"
          />
        </g>
      </g>
    </svg>
    <p
      class="uiLogo__logoName uiLogo__logoName--uiLogo"
    >
      Visual Poetry
    </p>
  </span>
</a>
`;
